providers/mailru/mailru.go

```diff
@@ -0,0 +1,166 @@
+// Package mailru implements the OAuth2 protocol for authenticating users through mailru.com.
+// This package can be used as a reference implementation of an OAuth2 provider for Goth.
+package mailru
+
+import (
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/markbates/goth"
+	"golang.org/x/oauth2"
+)
+
+const (
+	authURL      = "https://connect.mail.ru/oauth/authorize"
+	tokenURL     = "https://connect.mail.ru/oauth/token"
+	endpointUser = "https://www.appsmail.ru/platform/api"
+)
+
+// New creates a new MAILRU provider and sets up important connection details.
+// You should always call `mailru.New` to get a new provider. Never try to
+// create one manually.
+func New(clientID, clientSecret, redirectURL string, scopes ...string) *Provider {
+	var c = &oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		RedirectURL:  redirectURL,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  authURL,
+			TokenURL: tokenURL,
+		},
+		Scopes: []string{},
+	}
+
+	for _, scope := range scopes {
+		c.Scopes = append(c.Scopes, scope)
+	}
+
+	return &Provider{
+		name:        "mailru",
+		oauthConfig: c,
+	}
+}
+
+// Provider is the implementation of `goth.Provider` for accessing Github.
+type Provider struct {
+	name         string
+	clientID     string
+	clientSecret string
+	httpClient   *http.Client
+	oauthConfig  *oauth2.Config
+}
+
+// Name is the name used to retrieve this provider later.
+func (p *Provider) Name() string {
+	return p.name
+}
+
+// SetName is to update the name of the provider (needed in case of multiple providers of 1 type)
+func (p *Provider) SetName(name string) {
+	p.name = name
+}
+
+func (p *Provider) Client() *http.Client {
+	return goth.HTTPClientWithFallBack(p.httpClient)
+}
+
+// BeginAuth asks MAILRU for an authentication end-point.
+func (p *Provider) BeginAuth(state string) (goth.Session, error) {
+	return &Session{
+		AuthURL: p.oauthConfig.AuthCodeURL(state),
+	}, nil
+}
+
+// FetchUser will go to MAILRU and access basic information about the user.
+func (p *Provider) FetchUser(session goth.Session) (_ goth.User, err error) {
+	var (
+		sess = session.(*Session)
+		user = goth.User{
+			AccessToken:  sess.AccessToken,
+			RefreshToken: sess.RefreshToken,
+			Provider:     p.Name(),
+			ExpiresAt:    sess.ExpiresAt,
+		}
+	)
+
+	if user.AccessToken == "" {
+		return user, fmt.Errorf("%s cannot get user information without access token", p.name)
+	}
+
+	var builder strings.Builder
```

Comment:
`strings.Builder` is only available for Go 1.10 and later, but Goth supports back to Go 1.7. You'll need to either guard this with a build tag, or to make things easier, use some other means of creating this string. String concatenation is probably the way to go.

Reply:
Oops, fixed.